### PR TITLE
Include what is used in pybind11.h

### PR DESCRIPTION
**Pull request by Ricardicus**
Adding some includes that the file 'include/pybind11/pybind11.h' is depending on, but not including itself.

Add #include <memory> for unique_ptr<> (include/pybind11/pybind11.h:1132)
Add #include <vector> for vector<> (include/pybind11/pybind11.h:1735)
Add #include <string> for string  (include/pybind11/pybind11.h:2139)
Add #include <utility> for move  (include/pybind11/pybind11.h:2197)

### CODE REVIEW

File: include/pybind11/pybind11.h [hunk 1]
```diff
@@ -41,6 +41,11 @@
 #  endif
 #endif
 
+#include <memory>
+#include <vector>
+#include <string>
+#include <utility>
+
```
Comment (YannickJadoul) [inline include/pybind11/pybind11.h:48]: Before or after the other pybind11 headers? I tend to include stdlib as last includes, but I'm not sure if there's a policy or habit, here. Apart from that, fine with me! :-)

**Review by YannickJadoul — APPROVED** [review]

File: include/pybind11/pybind11.h (hunk 1 above)
Comment (Ricardicus) [reply include/pybind11/pybind11.h:48]: I can change the order, if you like? :)

Comment (Ricardicus) [reply include/pybind11/pybind11.h:48]: Here: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
I read: 

_Include headers in the following order: Related header, C system headers, C++ standard library headers, other libraries' headers, your project's headers._

But that is Googles rules, not necessarily yours.

Comment (henryiii) [reply include/pybind11/pybind11.h:48]: I think LLVM and therefore the `clang-*` tools have the same ordering recommendation.

Comment (henryiii) [reply include/pybind11/pybind11.h:48]: PS: There is one hard requirement for all Python modules, though: `Python.h` must precede all stdlib headers.

Comment (YannickJadoul) [reply include/pybind11/pybind11.h:48]: Also fine to keep it this way, then. Just wanted to open this for discussion, and check whether this is what we want :-)

**Comment by henryiii** [comment]
Why does this break appveyor? AFAICT, it's working on other branches, but not this, twice now?

File: include/pybind11/pybind11.h (hunk 1 above)
Comment (henryiii) [reply include/pybind11/pybind11.h:48]: Actually, I think this has to be moved. This puts stdlib includes above `<Python.h>`, which causes massive numbers of warnings on Python 2.7 on older Linux systems (at least). The rule is `Python.h` must be above any stdlib includes for any extension module, and that rule is kept by by pybind11.

**Comment by Ricardicus** [comment]
Interesting. I thought standard C++ library headers were safe to include anywhere. I have no experience of your pipeline, curious to know why it breaks.

File: include/pybind11/pybind11.h (hunk 1 above)
Comment (Ricardicus) [reply include/pybind11/pybind11.h:48]: Ok! I will do another commit soon, within the hour, moving them under the project headers!

**Comment by henryiii** [comment]
> I thought standard C++ library headers were safe to include anywhere.

It's a requirement listed in the CPython docs:

https://docs.python.org/3.8/c-api/intro.html#include-files

The easiest way I know of to trigger it doing something bad is to use CentOS 7 and the built-in Python 2. You get massive numbers of mismatched redefine warnings if you don't put Python.h first.

**Comment by Ricardicus** [comment]
Lets see if it fails, still.

**Comment by Ricardicus** [comment]
Looks like it went well 👍

**Review by henryiii — APPROVED** [review]

**Comment by YannickJadoul** [comment]
Thanks, @Ricardicus! :-)

**Comment by Ricardicus** [comment]
Thank you!

**Pull request merged**